File: backend/app/schemas/extraction.py
"""
Extraction Schemas.

Schemas Pydantic for extraction de data de articles cientificos.
"""

from datetime import datetime
from enum import Enum
from typing import Annotated, Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.core.config import settings

# =================== INTERNAL ASSEMBLY SCHEMAS ===================


class AssemblyInfo(BaseModel):
    """Internal (non-API) record for one prompt assembly — logged per run for
    token/cost observability and to surface windowing overflow.

    Deliberately NOT referenced by any endpoint response model, so it never
    enters the OpenAPI/schema.d.ts contract.
    """

    model_config = ConfigDict(frozen=True)

    total_blocks: int
    included_blocks: int
    truncated: bool
    est_tokens: int


# =================== COMMON SCHEMAS ===================


class ExtractionOptions(BaseModel):
    """Opcoes de extraction."""

    model: str = Field(default=settings.LLM_DEFAULT_MODEL, description="Model to use")
    temperature: float = Field(default=0.1, ge=0, le=2)
    max_tokens: int | None = Field(default=None, ge=100, le=16000)

    model_config = ConfigDict(populate_by_name=True)


class EvidencePassage(BaseModel):
    """Passagem de texto citada como evidencia."""

    text: str = Field(..., description="Texto extraido do documento")
    page_number: int | None = Field(default=None, description="Numero da pagina")
    confidence: float | None = Field(default=None, ge=0, le=1)

    model_config = ConfigDict(populate_by_name=True)


class FieldSuggestion(BaseModel):
    """Sugestao de valor for um field."""

    field_id: UUID = Field(..., alias="fieldId")
    field_name: str = Field(..., alias="fieldName")
    suggested_value: Any = Field(..., alias="suggestedValue")
    confidence_score: float | None = Field(default=None, alias="confidenceScore", ge=0, le=1)
    reasoning: str | None = None
    evidence: list[EvidencePassage] = []

    model_config = ConfigDict(populate_by_name=True)


# =================== SECTION EXTRACTION SCHEMAS ===================


class SectionExtractionRequest(BaseModel):
    """Request for extraction de section."""

    project_id: UUID = Field(..., alias="projectId")
    article_id: UUID = Field(..., alias="articleId")
    template_id: UUID = Field(..., alias="templateId")

    # Para extraction de section unica
    entity_type_id: UUID | None = Field(default=None, alias="entityTypeId")

    # Para extraction em batch de todas as sections
    parent_instance_id: UUID | None = Field(default=None, alias="parentInstanceId")
    extract_all_sections: bool = Field(default=False, alias="extractAllSections")

    # Filtrar sections especificas (para chunking)
    section_ids: list[UUID] | None = Field(default=None, alias="sectionIds")

    # Texto do PDF ja processado (evita reprocessar)
    pdf_text: str | None = Field(default=None, alias="pdfText")

    # Opcoes de extraction
    options: ExtractionOptions | None = None
    model: str | None = Field(
        default=settings.LLM_DEFAULT_MODEL,
        description="Model to use",
    )

    # Reuse an existing run instead of creating a new one. Used by the
    # Quality-Assessment surface, which opens a Run via the HITL session
    # service and then fires AI extraction against that same Run (so the
    # session-scoped Run isn't shadowed by a new one).
    run_id: UUID | None = Field(default=None, alias="runId")

    # Retained for API compatibility but inert in the collapsed lifecycle:
    # there is no separate ``review`` stage to auto-advance to, so the run
    # stays in ``extract`` after AI extraction and reviewers act there
    # directly. QA's publish flow walks the run from ``extract`` to
    # ``finalized`` itself. See ADR-0013 / section_extraction_service.
    auto_advance_to_review: bool = Field(default=True, alias="autoAdvanceToReview")

    # When True, fields a human has already settled are excluded from the
    # LLM call so users can re-run AI without losing their work. "Settled"
    # spans both tracks: a latest `human` proposal (QA surface) or a
    # committed reviewer decision — edit/accept_proposal (extraction
    # surface, where the blind-review gate routes human values to
    # ReviewerDecision rows rather than `human` proposals).
    skip_fields_with_human_proposals: bool = Field(
        default=False,
        alias="skipFieldsWithHumanProposals",
    )

    model_config = ConfigDict(populate_by_name=True)

    @model_validator(mode="after")
    def validate_extraction_mode(self) -> "SectionExtractionRequest":
        """Valida que os fields corretos estao presentes for cada modo."""
        # QA / pre-opened-run mode: caller passes ``run_id`` and the
        # service iterates every top-level entity_type of that run's
        # template. No parent_instance_id / entity_type_id required.
        if self.run_id is not None:
            return self
        if self.extract_all_sections:
            if not self.parent_instance_id:
                raise ValueError("parentInstanceId is required when extractAllSections is true")
        else:
            if not self.entity_type_id:
                raise ValueError("entityTypeId is required when extractAllSections is false")
        return self


class SectionOutcome(BaseModel):
    """Per-section outcome inside a batch extraction result.

    Wire format is snake_case on purpose — these items were emitted as
    raw service dicts before being typed, and the keys are preserved
    verbatim. Failure items carry ``error`` and omit the counters.
    """

    entity_type_id: str
    entity_type_name: str | None = None
    success: bool
    suggestions_created: int = 0
    tokens_used: int = 0
    skipped: bool = False
    error: str | None = None


class SingleSectionResult(BaseModel):
    """Resultado de extraction de section unica."""

    mode: Literal["single"] = "single"
    extraction_run_id: str = Field(..., alias="extractionRunId")
    suggestions_created: int = Field(..., alias="suggestionsCreated")
    entity_type_id: str = Field(..., alias="entityTypeId")
    tokens_prompt: int = Field(..., alias="tokensPrompt")
    tokens_completion: int = Field(..., alias="tokensCompletion")
    tokens_total: int = Field(..., alias="tokensTotal")
    duration_ms: float = Field(..., alias="durationMs")

    model_config = ConfigDict(populate_by_name=True)


class BatchSectionResult(BaseModel):
    """Resultado de extraction em batch."""

    mode: Literal["batch"] = "batch"
    extraction_run_id: str = Field(..., alias="extractionRunId")
    total_sections: int = Field(..., alias="totalSections")
    successful_sections: int = Field(..., alias="successfulSections")
    failed_sections: int = Field(..., alias="failedSections")
    total_suggestions_created: int = Field(..., alias="totalSuggestionsCreated")
    total_tokens_used: int = Field(..., alias="totalTokensUsed")
    duration_ms: float = Field(..., alias="durationMs")
    sections: list[SectionOutcome] = Field(default=[], alias="sections")

    model_config = ConfigDict(populate_by_name=True)


# Discriminated union: the ``mode`` tag lets OpenAPI consumers (and the
# generated TypeScript types) narrow the payload instead of guessing by
# field presence.
SectionExtractionResponseData = Annotated[
    SingleSectionResult | BatchSectionResult,
    Field(discriminator="mode"),
]


# =================== MODEL EXTRACTION SCHEMAS ===================


class CreateModelHierarchyRequest(BaseModel):
    """Request to create one prediction-model hierarchy for an article."""

    project_id: UUID = Field(..., alias="projectId")
    article_id: UUID = Field(..., alias="articleId")
    template_id: UUID = Field(..., alias="templateId")
    model_name: str = Field(..., alias="modelName")
    modelling_method: str | None = Field(default=None, alias="modellingMethod")

    model_config = ConfigDict(populate_by_name=True)


class ModelHierarchyChildResponse(BaseModel):
    """Child instance created under the parent model instance."""

    id: UUID
    entity_type_id: UUID = Field(..., alias="entityTypeId")
    parent_instance_id: UUID = Field(..., alias="parentInstanceId")
    label: str

    model_config = ConfigDict(populate_by_name=True)


class CreateModelHierarchyResponse(BaseModel):
    """Response for one-shot hierarchy creation."""

    model_id: UUID = Field(..., alias="modelId")
    model_label: str = Field(..., alias="modelLabel")
    child_instances: list[ModelHierarchyChildResponse] = Field(alias="childInstances")
    proposal_run_id: UUID | None = Field(default=None, alias="proposalRunId")

    model_config = ConfigDict(populate_by_name=True)


class ModelExtractionRequest(BaseModel):
    """Request for extraction de modelos de predicao."""

    project_id: UUID = Field(..., alias="projectId")
    article_id: UUID = Field(..., alias="articleId")
    template_id: UUID = Field(..., alias="templateId")

    # Opcoes de extraction
    model: str | None = Field(
        default=settings.LLM_DEFAULT_MODEL,
        description="Model to use",
    )
    options: ExtractionOptions | None = None

    model_config = ConfigDict(populate_by_name=True)


class IdentifiedModel(BaseModel):
    """Modelo de predicao identificado in the article."""

    model_name: str = Field(..., alias="modelName")
    model_type: str | None = Field(default=None, alias="modelType")
    target_outcome: str | None = Field(default=None, alias="targetOutcome")
    description: str | None = None

    # Metadata adicionais extraidos
    sample_size: int | None = Field(default=None, alias="sampleSize")
    performance_metrics: dict[str, Any] = Field(default={}, alias="performanceMetrics")
    validation_strategy: str | None = Field(default=None, alias="validationStrategy")

    model_config = ConfigDict(populate_by_name=True)


class CreatedModelInfo(BaseModel):
    """One prediction-model instance created by model extraction."""

    instance_id: str = Field(..., alias="instanceId")
    model_name: str = Field(..., alias="modelName")
    modelling_method: str | None = Field(default=None, alias="modellingMethod")

    model_config = ConfigDict(populate_by_name=True)


class ModelExtractionRunStats(BaseModel):
    """Timing/token metadata attached to a model-extraction response."""

    duration: int
    models_found: int = Field(..., alias="modelsFound")
    tokens_prompt: int = Field(..., alias="tokensPrompt")
    tokens_completion: int = Field(..., alias="tokensCompletion")
    tokens_total: int = Field(..., alias="tokensTotal")

    model_config = ConfigDict(populate_by_name=True)


class ModelExtractionResult(BaseModel):
    """Resultado da extraction de modelos."""

    extraction_run_id: str = Field(..., alias="extractionRunId")
    models_created: list[CreatedModelInfo] = Field(..., alias="modelsCreated")
    total_models: int = Field(..., alias="totalModels")
    child_instances_created: int = Field(..., alias="childInstancesCreated")
    metadata: ModelExtractionRunStats

    model_config = ConfigDict(populate_by_name=True)


# =================== ENTITY TYPE SCHEMAS ===================


class ExtractionFieldSchema(BaseModel):
    """Schema de field de extraction."""

    id: UUID
    name: str
    label: str
    description: str | None = None
    field_type: str = Field(..., alias="fieldType")
    is_required: bool = Field(default=False, alias="isRequired")
    allowed_values: list[str] | None = Field(default=None, alias="allowedValues")
    unit: str | None = None
    allowed_units: list[str] | None = Field(default=None, alias="allowedUnits")
    llm_description: str | None = Field(default=None, alias="llmDescription")
    sort_order: int = Field(default=0, alias="sortOrder")

    model_config = ConfigDict(populate_by_name=True)


class ExtractionEntityTypeSchema(BaseModel):
    """Schema de tipo de entidade."""

    id: UUID
    name: str
    label: str
    description: str | None = None
    cardinality: Literal["one", "many"]
    # ``role`` is the structural discriminant the frontend partitions on
    # (study-level accordions vs model container vs per-model children).
    # See ``ExtractionEntityRole`` and migration ``0016_entity_role_column``.
    role: Literal["study_section", "model_container", "model_section"] = "study_section"
    is_required: bool = Field(default=False, alias="isRequired")
    sort_order: int = Field(default=0, alias="sortOrder")
    parent_entity_type_id: UUID | None = Field(default=None, alias="parentEntityTypeId")
    fields: list[ExtractionFieldSchema] = []

    model_config = ConfigDict(populate_by_name=True)


class ExtractionTemplateSchema(BaseModel):
    """Schema de template de extraction."""

    id: UUID
    name: str
    description: str | None = None
    framework: Literal["CHARMS", "PICOS", "CUSTOM"]
    version: str
    entity_types: list[ExtractionEntityTypeSchema] = Field(default=[], alias="entityTypes")

    model_config = ConfigDict(populate_by_name=True)


# =================== INSTANCE SCHEMAS ===================


class CreateInstanceRequest(BaseModel):
    """Request for criar instance de extraction."""

    project_id: UUID = Field(..., alias="projectId")
    article_id: UUID = Field(..., alias="articleId")
    template_id: UUID = Field(..., alias="templateId")
    entity_type_id: UUID = Field(..., alias="entityTypeId")
    parent_instance_id: UUID | None = Field(default=None, alias="parentInstanceId")
    label: str
    metadata: dict[str, Any] = {}

    model_config = ConfigDict(populate_by_name=True)


class InstanceResponse(BaseModel):
    """Response de instance de extraction."""

    id: UUID
    project_id: UUID = Field(..., alias="projectId")
    article_id: UUID | None = Field(default=None, alias="articleId")
    template_id: UUID = Field(..., alias="templateId")
    entity_type_id: UUID = Field(..., alias="entityTypeId")
    parent_instance_id: UUID | None = Field(default=None, alias="parentInstanceId")
    label: str
    sort_order: int = Field(..., alias="sortOrder")
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


# =================== VALUE SCHEMAS ===================


class SaveValueRequest(BaseModel):
    """Request for salvar valor extraido."""

    instance_id: UUID = Field(..., alias="instanceId")
    field_id: UUID = Field(..., alias="fieldId")
    value: Any
    source: Literal["human", "ai", "rule"] = "human"
    unit: str | None = None
    evidence: list[EvidencePassage] = []

    model_config = ConfigDict(populate_by_name=True)


class ValueResponse(BaseModel):
    """Response de valor extraido."""

    id: UUID
    instance_id: UUID = Field(..., alias="instanceId")
    field_id: UUID = Field(..., alias="fieldId")
    value: Any
    source: str
    confidence_score: float | None = Field(default=None, alias="confidenceScore")
    unit: str | None = None
    is_consensus: bool = Field(default=False, alias="isConsensus")
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


# =================== SUGGESTION SCHEMAS ===================


class SuggestionResponse(BaseModel):
    """Response de suggestion de IA for extraction."""

    id: UUID
    extraction_run_id: UUID = Field(..., alias="extractionRunId")
    instance_id: UUID | None = Field(default=None, alias="instanceId")
    field_id: UUID = Field(..., alias="fieldId")
    suggested_value: Any = Field(..., alias="suggestedValue")
    confidence_score: float | None = Field(default=None, alias="confidenceScore")
    reasoning: str | None = None
    status: Literal["pending", "accepted", "rejected"]
    created_at: datetime = Field(..., alias="createdAt")

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


class ReviewSuggestionRequest(BaseModel):
    """Request for revisar suggestion."""

    status: Literal["accepted", "rejected"]
    modified_value: Any | None = Field(default=None, alias="modifiedValue")

    model_config = ConfigDict(populate_by_name=True)


# =================== ASYNC SECTION EXTRACTION JOB SCHEMAS ===================


class ExtractionErrorCode(str, Enum):
    """Stable, machine-readable code for a terminal extraction failure.

    Carried on ``ExtractionJobStatusResponse.error_code`` so the frontend can
    pick specific, actionable toast copy without parsing the human ``error``
    string. Only modes the backend genuinely raises are represented — each
    value has a real emitter in ``classify_extraction_error``:

    - ``PDF_NOT_FOUND``    — the article has no stored PDF (``FileNotFoundError``).
    - ``MISSING_API_KEY``  — no usable LLM key, BYOK or global (``MissingLLMKeyError``).
    - ``EXTRACTION_FAILED``— generic catch-all for everything else.
    """

    PDF_NOT_FOUND = "PDF_NOT_FOUND"
    MISSING_API_KEY = "MISSING_API_KEY"
    EXTRACTION_FAILED = "EXTRACTION_FAILED"


class ExtractionJobStartedResponse(BaseModel):
    """202 payload returned when an extraction job is queued."""

    job_id: str
    message: str | None = "Extraction queued. Poll /status/{job_id} for result."


class ExtractionJobResult(BaseModel):
    """Normalised result from a completed extraction job.

    Both single-section and batch result shapes are represented; fields
    absent for the other mode are ``None``.
    """

    model_config = ConfigDict(populate_by_name=True)

    mode: str
    extraction_run_id: str = Field(..., alias="extractionRunId")
    # Single-section fields
    suggestions_created: int | None = Field(default=None, alias="suggestionsCreated")
    entity_type_id: str | None = Field(default=None, alias="entityTypeId")
    # Batch fields
    total_sections: int | None = Field(default=None, alias="totalSections")
    successful_sections: int | None = Field(default=None, alias="successfulSections")
    failed_sections: int | None = Field(default=None, alias="failedSections")
    total_suggestions_created: int | None = Field(default=None, alias="totalSuggestionsCreated")
    sections: list[SectionOutcome] | None = Field(default=None, alias="sections")


class ExtractionJobStatusResponse(BaseModel):
    """GET /extraction/sections/status/{job_id} payload."""

    model_config = ConfigDict(populate_by_name=True)

    job_id: str = Field(..., alias="jobId")
    status: str = Field(..., description="pending | running | completed | failed | cancelled")
    result: ExtractionJobResult | None = None
    error: str | None = None
    # Machine-readable failure classification, set only when ``status`` is
    # ``failed``. Lets the frontend map to specific copy without parsing
    # ``error``. See ``ExtractionErrorCode`` / ``classify_extraction_error``.
    error_code: ExtractionErrorCode | None = Field(default=None, alias="errorCode")


# =================== CITATION ANCHOR (extraction_evidence.position v1) ===================
#
# Wire format for ``extraction_evidence.position`` JSONB. Mirrors
# ``frontend/pdf-viewer/core/citation.ts`` and ``coordinates.ts`` field-for-
# field, including camelCase JSON keys, so the viewer can read citation
# rows from the API and render them without translation.
#
# Spec: ``docs/superpowers/specs/2026-04-28-pdf-viewer-database-requirements.md``
# (sections 2 + 3).


class PDFRect(BaseModel):
    """Bounding box in PDF user space (origin bottom-left, points).

    Mirrors ``frontend/pdf-viewer/core/coordinates.ts:14-19``.
    """

    x: float
    y: float
    width: float
    height: float

    model_config = ConfigDict(populate_by_name=True)


class PDFTextRange(BaseModel):
    """A range of characters within a single page's text content.

    ``charStart`` / ``charEnd`` are offsets into the page's concatenated
    text — same offsets emitted by the engine's ``getTextContent()``.
    Mirrors ``frontend/pdf-viewer/core/coordinates.ts:26-30``.
    """

    page: int = Field(..., ge=1, description="1-indexed page number")
    char_start: int = Field(..., alias="charStart", ge=0)
    char_end: int = Field(..., alias="charEnd", ge=0)

    model_config = ConfigDict(populate_by_name=True)

    @model_validator(mode="after")
    def _char_range_valid(self) -> "PDFTextRange":
        if self.char_end < self.char_start:
            raise ValueError("charEnd must be >= charStart")
        return self


class TextCitationAnchor(BaseModel):
    """Char-range only — most robust to re-OCR / PDF re-encoding."""

    kind: Literal["text"]
    range: PDFTextRange
    quote: str | None = Field(
        default=None,
        description="Optional canonical text used for highlight matching",
    )
    block_ids: list[int] = Field(
        default_factory=list,
        alias="blockIds",
        description="block_index values (per page) the quote matched; reader highlight key",
    )

    model_config = ConfigDict(populate_by_name=True)


class RegionCitationAnchor(BaseModel):
    """Bbox only — works for figures, tables, image regions."""

    kind: Literal["region"]
    page: int = Field(..., ge=1, description="1-indexed page the rect is on")
    rect: PDFRect
    block_ids: list[int] = Field(
        default_factory=list,
        alias="blockIds",
        description="block_index values (per page) the quote matched; reader highlight key",
    )

    model_config = ConfigDict(populate_by_name=True)


class HybridCitationAnchor(BaseModel):
    """Both range and bbox + canonical quote — recommended for AI citations."""

    kind: Literal["hybrid"]
    range: PDFTextRange
    rect: PDFRect
    quote: str
    block_ids: list[int] = Field(
        default_factory=list,
        alias="blockIds",
        description="block_index values (per page) the quote matched; reader highlight key",
    )

    model_config = ConfigDict(populate_by_name=True)


CitationAnchor = Annotated[
    TextCitationAnchor | RegionCitationAnchor | HybridCitationAnchor,
    Field(discriminator="kind"),
]


class PositionV1(BaseModel):
    """v1 wire format for ``extraction_evidence.position`` JSONB.

    Discriminated on ``anchor.kind``; service-layer writers must populate
    ``extraction_evidence.text_content`` and ``page_number`` to match the
    anchor (see spec §2).
    """

    version: Literal[1]
    anchor: CitationAnchor

    model_config = ConfigDict(populate_by_name=True)


def parse_position(
    raw: dict[str, Any] | None,
) -> PositionV1 | None:
    """Validate a ``extraction_evidence.position`` JSONB payload.

    Accepts ``None`` (legacy rows with no position attached) or a dict in
    the v1 shape. Anything in between raises ``pydantic.ValidationError``
    so callers can surface a clear contract violation.
    """

    if raw is None:
        return None
    return PositionV1.model_validate(raw)

File: frontend/types/extraction.ts
/**
 * TypeScript types for the data extraction module
 *
 * This file defines all interfaces and types required for the
 * structured data extraction system.
 *
 * Includes Zod schemas for runtime validation.
 */

import {z} from 'zod';

// =================== ENUMS ===================

type ExtractionFramework = 'CHARMS' | 'PICOS' | 'CUSTOM';
export type ExtractionFieldType = 'text' | 'number' | 'date' | 'select' | 'multiselect' | 'boolean';
export type ExtractionCardinality = 'one' | 'many';
/**
 * Structural role of an entity type within a template.
 *
 * Mirrors the backend ``ExtractionEntityRole`` enum (migration
 * ``0016_entity_role_column``). Replaces the legacy convention of
 * identifying the model container by ``name === 'prediction_models'``.
 */
export type ExtractionEntityRole = 'study_section' | 'model_container' | 'model_section';

/**
 * Extraction value type by field type
 * Ensures type safety instead of using `any`
 */
export type ExtractionValue = 
  | string      // text, select
  | number      // number
  | Date        // date
  | string[]    // multiselect
  | boolean     // boolean
    | null;       // unfilled values

// =================== TEMPLATES ===================

/**
 * Standardized global template (CHARMS, PICOS, PRISMA, etc.)
 * Maintained by admins, read-only for users
 */
export interface GlobalExtractionTemplate {
  id: string;
  name: string;
  description: string | null;
  framework: ExtractionFramework;
  version: string;
  is_global: boolean;
  schema: any;
  created_at: string;
  updated_at: string;
}

export interface ProjectExtractionTemplate {
  id: string;
  project_id: string;
  global_template_id: string | null;
  name: string;
  description: string | null;
  framework: ExtractionFramework;
  version: string;
  schema: any;
  is_active: boolean;
  created_by: string;
  created_at: string;
  updated_at: string;
}

// =================== ENTIDADES E CAMPOS ===================

export interface ExtractionEntityType {
  id: string;
  template_id: string;
  name: string;
  label: string;
  description: string | null;
  parent_entity_type_id: string | null;
  cardinality: ExtractionCardinality;
  /**
   * Structural discriminant the UI partitions on. See
   * ``partitionEntityTypes`` and ``isModelContainer``.
   */
  role: ExtractionEntityRole;
  sort_order: number;
  is_required: boolean;
  /**
   * Repeating-group entry noun (B-8): set only on ``model_container``
   * rows; consumers interpolate it with a ``'model'`` fallback.
   * Required (not optional) so hand-written mirrors and adapters cannot
   * silently drop it.
   */
  entry_label: string | null;
  created_at: string;
}

/**
 * Entity type joined with its fields list — the shape the form needs to
 * render. Lives at the type layer instead of being redeclared in every
 * consumer.
 */
export interface ExtractionEntityTypeWithFields extends ExtractionEntityType {
  fields: ExtractionField[];
}

export interface ExtractionField {
  id: string;
  entity_type_id: string;
  name: string;
  label: string;
  description: string | null;
  field_type: ExtractionFieldType;
  is_required: boolean;
  validation_schema: any;
  allowed_values: string[] | null;
  unit: string | null;
    allowed_units: string[] | null; // Configurable alternative units for number fields
    llm_description: string | null; // Specific instruction for AI extraction
  sort_order: number;
  created_at: string;
    // "Other (specify)" support inline
  allow_other?: boolean;
  other_label?: string | null;
  other_placeholder?: string | null;
    // ADR-0016 opt-in dispositions (no_information is universal, needs no flag)
  allows_not_applicable?: boolean;
  allows_not_evaluated?: boolean;
}

// =================== INSTANCES AND VALUES ===================

export interface ExtractionInstance {
  id: string;
  project_id: string;
  article_id: string;
  template_id: string;
  entity_type_id: string;
  parent_instance_id: string | null;
  label: string;
  sort_order: number;
  metadata: any;
  created_by: string;
  created_at: string;
  updated_at: string;
}


// =================== ZOD SCHEMAS (Runtime validation) ===================

/**
 * Zod schema for extraction field validation
 * Used when creating or editing fields
 */
export const ExtractionFieldSchema = z.object({
  name: z.string()
      .regex(/^[a-z][a-z0-9_]*$/, 'Name must be in snake_case (e.g. field_example)')
      .min(2, 'Name must be at least 2 characters')
      .max(50, 'Name must be at most 50 characters'),
  
  label: z.string()
      .min(1, 'Label is required')
      .max(100, 'Label must be at most 100 characters'),
  
  description: z.string()
      .max(500, 'Description must be at most 500 characters')
    .optional()
    .nullable(),
  
  field_type: z.enum(['text', 'number', 'date', 'select', 'multiselect', 'boolean'], {
      errorMap: () => ({message: 'Invalid field type'}),
  }),
  
  is_required: z.boolean().default(false),
  
  unit: z.string()
      .max(50, 'Unit must be at most 50 characters')
    .optional()
    .nullable(),
  
  allowed_units: z.array(z.string().max(50))
      .min(1, 'Must have at least one alternative unit')
      .max(20, 'Maximum of 20 alternative units')
    .optional()
    .nullable()
    .refine(
      (units) => {
        if (!units) return true;
        const unique = new Set(units);
        return unique.size === units.length;
      },
        {message: 'Units cannot have duplicates'}
    ),
  
  llm_description: z.string()
      .max(1000, 'AI instruction must be at most 1000 characters')
    .optional()
    .nullable(),
  
  allowed_values: z.array(z.string())
      .min(1, 'Must have at least one allowed value')
      .max(100, 'Maximum of 100 allowed values')
    .optional()
    .nullable()
    .refine(
      (values) => {
        if (!values) return true;
        const unique = new Set(values);
        return unique.size === values.length;
      },
        {message: 'Allowed values cannot have duplicates'}
    ),

  // "Other (specify)" support
  allow_other: z.boolean().default(false).optional(),
  // No default: a .default() wrapped in .optional() never fires (dead
  // code, removed at B-7) — the runtime fallback is the English copy key.
  other_label: z.string()
      .max(100, '"Other" label must be at most 100 characters')
    .optional()
    .nullable(),
  other_placeholder: z.string()
      .max(200, 'Placeholder must be at most 200 characters')
    .optional()
    .nullable(),

  // ADR-0016 opt-in disposition flags (no_information is universal, needs none)
  allows_not_applicable: z.boolean().default(false).optional(),
  allows_not_evaluated: z.boolean().default(false).optional(),

  validation_schema: z.record(z.any())
    .optional()
    .nullable(),
  
  sort_order: z.number()
      .int('Order must be an integer')
      .min(0, 'Order must be greater than or equal to 0')
    .default(0),
});

/**
 * Inferred type from Zod schema
 */
type ExtractionFieldInput = z.infer<typeof ExtractionFieldSchema>;

/**
 * Partial type for update (all fields optional)
 */
export type ExtractionFieldUpdate = Partial<ExtractionFieldInput>;

/**
 * Type for DB insert (adds entity_type_id)
 */
export interface ExtractionFieldInsert extends Omit<ExtractionFieldInput, 'sort_order'> {
  entity_type_id: string;
  sort_order?: number;
  allowed_units?: string[] | null;
}

// =================== FIELD MANAGEMENT TYPES ===================

// FieldValidationResult lives in services/extractionFieldService.ts, next to
// the call that produces it. The copy that used to sit here was a stale
// duplicate (message optional instead of required) that only tests reached.


/**
 * User role in the project (for permission control)
 */
export type ProjectMemberRole = 'manager' | 'reviewer' | 'viewer' | 'consensus';

/**
 * Result of permission check
 */
export interface PermissionCheckResult {
  canView: boolean;
  canEdit: boolean;
  canDelete: boolean;
  canCreate: boolean;
  role: ProjectMemberRole | null;
  message?: string;
}

